cdap-app-templates/cdap-etl/cdap-data-pipeline/src/test/java/co/cask/cdap/datapipeline/LineageOperationProcessorTest.java

```diff
@@ -0,0 +1,287 @@
+/*
+ * Copyright © 2018 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.datapipeline;
+
+import co.cask.cdap.api.lineage.field.EndPoint;
+import co.cask.cdap.api.lineage.field.InputField;
+import co.cask.cdap.data2.metadata.lineage.field.FieldLineageInfo;
+import co.cask.cdap.etl.api.lineage.field.Operation;
+import co.cask.cdap.etl.api.lineage.field.ReadOperation;
+import co.cask.cdap.etl.api.lineage.field.TransformOperation;
+import co.cask.cdap.etl.api.lineage.field.WriteOperation;
+import co.cask.cdap.etl.proto.Connection;
+import org.junit.Assert;
+import org.junit.Test;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ *
+ */
+public class LineageOperationProcessorTest {
+
+  @Test
```

Comment:
the test cases here make sense, but I am not sure they cover all cases.
What about a pipeline where:
- a stage has two inputs
- both of these inputs have "pass-through" fields (that are not touched by that transform but come from the previous stage)
- that means you have a merge that needs to output fields that are not direct outputs of the preceding stages. 

Also wondering what a "rename" would do. It essentially drops a field and creates a new one. How is that represented, and how is that handled when finding the origins of an operation's input?

Reply:
Added test case with complex merge and rename scenario - `testComplexMerge`